Fix TOD incorrectly determining state between sun events

jorhett opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change

Remove a conditional from #65884 so that the "between" code is correctly executed whether it's a sun event or a fixed time. This attempts to fix a bug where Time of Day is wrong if HA is restarted after midnight between sunset and sunrise. 

## Type of change

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #70206 #69288 #67039
- This PR is related to issue: #70206
- Link to documentation pull request: https://www.home-assistant.io/integrations/tod/

## Checklist

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

homeassistant commented:
Hi @jorhett, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#70572).

Once you do that we will be able to review and accept this pull request.

Thanks!

jorhett commented:
I was going to modify the tests to catch this, but honestly I was confused as to why the current tests didn't catch this bug. My guess is that the freezer time moves don't correctly simulate an HA restart. It probably just needs to re-run the same tests after simulating a restart.

nvx commented:
Have you actually tested the commit? Looks like it's got a syntax error to me.

The check for not being a sun event is important in this case as there are edge cases where the sun may not rise and/or set in a particular 24h period when in some locations like the arctic circle. Removing that check breaks a number of unit tests that check for correct behaviour in those cases.

I'd definitely try and reproduce the issue with a unit test that fails under the current code first so you can better understand the issue before trying to fix it. The issue I fixed (which was time of day based rather than sun position based) took a bit of thinking to properly understand what was going on and I had the benefit of a prior patch for the same issue to reference. Time is painful!

bdraco commented:
Given how many iterations there have been to try to get this right, we need a test for this before we can merge

emontnemery commented:
I'm marking this as draft for now, @jorhett please mark as ready for review when you've added a test which fails without your test and passes with it and also tested your patch in a running Home Assistant 👍

jorhett commented:
> Have you actually tested the commit? Looks like it's got a syntax error to me.
> 
> The check for not being a sun event is important in this case 

This response was pretty snarky, especially given that one **would think** it would be important in PR #65884 that says it addresses Sunset/Sunrise and yet that PR includes this if condition to ignore them. Between you and I, my PR was mostly intended as a prompt to address what that PR title says it addressed but didn't.

My inability to create a test is not a lack of competence in Python. I work in Python all day, every day in my day job. But the instructions for testing always refer to running stuff locally which isn't relevant or of interest to me, and is likely how such problems creep in since local environments are always impacted by local config decisions outside of the code. 

I'm willing to learn, but I'm going to need instructions for building isolated test suites and documentation for all the wrappers you use. Wrappers (like the freezer) appear to be completely undocumented, so asking an outsider to build tests using the undocumented wrappers is basically sending the outsider down a rabbit hole to learn the entire code base.

TL;DR I'd like to help, I don't need Python clue but I do need a coherent, understandable testing methodology.

jorhett commented:
> Given how many iterations there have been to try to get this right, we need a test for this before we can merge

When I read the test suite, this situation is covered. Yet the tests are not catching it, which points to a problem in the testing implementation not the lack of tests. See my other comment about setting up a working test environment

nvx commented:
> > Have you actually tested the commit? Looks like it's got a syntax error to me.
> > The check for not being a sun event is important in this case
> 
> This response was pretty snarky, especially given that one **would think** it would be important in PR #65884 that says it addresses Sunset/Sunrise and yet that PR includes this if condition to ignore them. Between you and I, my PR was mostly intended as a prompt to address what that PR title says it addressed but didn't.

That PR addresses time of day (specifically when specified as a time, not when specified as a sunrise/sunset offset) being calculated incorrectly in certain edge cases between sunrise/sunset when home assistant starts. The title of the PR is actually copied from the original fix (which was inadvertently reverted later on) which I re-applied and added a unit test to prevent such accidental reverts in the past.

It is a little confusing as the tod (time of day) component can be configured with an actual time (which is what that PR addresses), but also can be configured as a sunrise or sunset offset (which is what this issue is about).

The specific case fixed by that PR is best demonstrated in the unit test added as a part of it:
https://github.com/home-assistant/core/pull/65884/files#diff-a62f235794d15195af8368d3ee7307a79c64b2f2e0c4d808aba4129ecb06bf73
The edge case in question there configures tod as "before: 12:00 after: 23:00" which at 10am should be considered on, while in practice it was being considered as off when Home Assistant first starts if home assistant was started between midnight and midday (home assistant starting after 23:00 would calculate correctly)

> My inability to create a test is not a lack of competence in Python. I work in Python all day, every day in my day job. But the instructions for testing always refer to running stuff locally which isn't relevant or of interest to me, and is likely how such problems creep in since local environments are always impacted by local config decisions outside of the code.

I was not implying anything about your competence in Python one way or the other. I was simply pointing out what appeared to be a simple syntax error, along with a note that simply removing that is_sun_event condition would break other sun edge case conditions which are currently (correctly) checked by unit tests.

My understanding of the problem I was fixing seemed at the time to only affect when an absolute time was specified rather than a sunrise/sunset offset as there is other logic that handles special edge cases for sunrise/sunset offsets separately, and indeed applying the same logic to sunrise/sunset offsets broke some of that special handling so I specifically excluded it.

The complexity is that "10am" will always happen every day everywhere in the world, while sunrise or sunset (and thus any offset from either times) may or may not happen every day in some parts of the world. My fix ensures that we don't accidentally wait until tomorrow to turn on the state if (using the test example above) 23:00 has already passed (which may have been yesterday) and we haven't yet reached the "before" time. This simple logic falls apart if the given time doesn't happen every day however. I expect the fix will involve much more complex logic than #65884 has that is specific to sun events.

> I'm willing to learn, but I'm going to need instructions for building isolated test suites and documentation for all the wrappers you use. Wrappers (like the freezer) appear to be completely undocumented, so asking an outsider to build tests using the undocumented wrappers is basically sending the outsider down a rabbit hole to learn the entire code base.

Definitely agreed on this one, #65884 was my first PR to home assistant and I struggled to get the environment setup to be able to run unit tests initially as well. The docs at https://developers.home-assistant.io/docs/development_environment could do with being expanded a little to make it easier to get started.

> TL;DR I'd like to help, I don't need Python clue but I do need a coherent, understandable testing methodology.

nvx commented:
> > Given how many iterations there have been to try to get this right, we need a test for this before we can merge
> 
> When I read the test suite, this situation is covered. Yet the tests are not catching it, which points to a problem in the testing implementation not the lack of tests. See my other comment about setting up a working test environment

While I haven't used tod with sun events enough to run into this issue, my guess is if it's a similar issue to the PR I fixed it needs some very specific conditions to trigger which may not be the case in the existing unit tests.

* after must be later in the day than "before", ie after: sunset, before: sunrise. There may be even more edge cases where using a mix of sun events and absolute times.
* home assistant must "start" when you're expecting the state to be on, but in the day after the "after" condition (but still before the "before" time)

binary_sensor:
  - platform: tod
    name: sundown
    after: sunset
    before: sunrise

Testing that for example at 1am in a location near the equator I would expect would trigger the issue.

I expect eg test_from_sunset_to_sunrise is not catching this because it is looking up the next sunrise and faking the time as that, so we're effectively in the "one day has passed since restarting home assistant" case there which is when it works.

I'd write a unit test that specifically hardcodes in times and dates similar to my unit test along with a location near the equator to test the "every day" sort of case, as well as another testing the arctic circle edge case (where sunset was 2 days ago and sunset isn't until another 2 days from now or the like).

jorhett commented:
> applying the same logic to sunrise/sunset offsets broke some of that special handling so I specifically excluded it.

This removed sun time support from the only component that supports a period defined by it

> The complexity is that "10am" will always happen every day everywhere in the world, while sunrise or sunset (and thus any offset from either times) may or may not happen every day in some parts of the world. 

I'm not actually seeing that it works at all. I completely understand why you did it and how your fix works for something other than what the title of your PR and also what the HA ChangeLog said it fixed. The confusing part is that a small fraction of a single percent of users who don't have both a sunset and a sunrise in a 24-hour period might be impacted feels very odd to me. Is there even a single HA user today that currently fits that description?  Even if say there were 900 of them (which I doubt) wouldn't fixing it for the (I'm sure) greater than 90,000 users with a note that we'll fix it for the small percentage later be a useful increment forward?

> This simple logic falls apart if the given time doesn't happen every day however.

Again, it's broken for everyone. Why not fix it for 99% now and iterate for the small percentage later?

And to be clear again: I'm not against fixing it in full. But you're not using normal Python datetime, it's all a bunch of under- (if any) documented special components. If you want people to jump in and help, it needs to not involve a man-month of learning the entire codebase from scratch.

jorhett commented:
> > When I read the test suite, this situation is covered. Yet the tests are not catching it, which points to a problem in the testing implementation not the lack of tests. See my other comment about setting up a working test environment
 
> While I haven't used tod with sun events enough to run into this issue, my guess is if it's a similar issue to the PR I fixed it needs some very specific conditions to trigger which may not be the case in the existing unit tests.
> 
> * after must be later in the day than "before", ie after: sunset, before: sunrise. There may be even more edge cases where using a mix of sun events and absolute times.

The current unit tests include one minute after sunset and one minute before sunrise. Both of those should be within the period. Yet in reality the latter time does not work.  The test is incorrectly written (using an undocumented mechanism that I gave up trying to figure out)

> * home assistant must "start" when you're expecting the state to be on, but in the day after the "after" condition (but still before the "before" time)

AND after midnight. It works fine if it restarts before midnight. 

That is the main problem with this code -- it's way too "what day am I in?" oriented.  Sunset happens at a time. Sunrise happens at a time. Times are integers so it's all simple math. This code needs to stop thinking about days.

> Testing that for example at 1am in a location near the equator I would expect would trigger the issue.

Testing at 1am ANYWHERE triggers the issue. 

> I expect eg test_from_sunset_to_sunrise is not catching this because it is looking up the next sunrise and faking the time as that, so we're effectively in the "one day has passed since restarting home assistant" case there which is when it works.

The problem is identical to the one you solved. After a restart the sunset event is in the future, not in the past. Forget about days and reset to the sunset prior to the restart.

> I'd write a unit test that specifically hardcodes in times and dates

Honestly I think that the current code is thinking way too much about specific conditionals around date, location, etc. None of that matters. Reduce Start to an integer, End to an integer, and compare simple math. Anything else is distracting in the code and misleading the reader.  With simple integers, it doesn't matter if there are 72 years between sunset and sunrise on a moon of Saturn or whatever, none of the code needs to change 😁

jorhett commented:
> also tested your patch in a running Home Assistant setup 👍

by running on my desktop where my personal configs impact how it operates?  Both of your documented test environments at https://developers.home-assistant.io/docs/development_environment/ are local installations within the developer OS. That is yet another way to waste development time dealing with issues that have no bearing on the code I am fixing, targeted for HAOS which has nothing in common with my desktop environment and completely irrelevant to the IDE I am using

I'm truly not trying to be a pain here. I do want to test. I love tests. But I'm way too experienced to waste time chasing down rabbit holes that always end up in local OS differences from the target platform.

nvx commented:
> > applying the same logic to sunrise/sunset offsets broke some of that special handling so I specifically excluded it.
> 
> This removed sun time support from the only component that supports a period defined by it

I'm not sure I follow. That PR made a change to the behaviour of the tod component in the specific case when you're not using sunrise/sunset. Behaviour of a configuration using tod with sunrise/sunset is unchanged if that PR is applied or not.

> > The complexity is that "10am" will always happen every day everywhere in the world, while sunrise or sunset (and thus any offset from either times) may or may not happen every day in some parts of the world.
> 
> I'm not actually seeing that it works at all. I completely understand why you did it and how your fix works for something other than what the title of your PR and also what the HA ChangeLog said it fixed. The confusing part is that a small fraction of a single percent of users who don't have both a sunset and a sunrise in a 24-hour period might be impacted feels very odd to me. Is there even a single HA user today that currently fits that description? Even if say there were 900 of them (which I doubt) wouldn't fixing it for the (I'm sure) greater than 90,000 users with a note that we'll fix it for the small percentage later be a useful increment forward?

Again, my PR fixed behaviour when not using sunrise/sunset, and did not change any behaviour when using sunrise/sunset.

> > This simple logic falls apart if the given time doesn't happen every day however.
> 
> Again, it's broken for everyone. Why not fix it for 99% now and iterate for the small percentage later?
> 
> And to be clear again: I'm not against fixing it in full. But you're not using normal Python datetime, it's all a bunch of under- (if any) documented special components. If you want people to jump in and help, it needs to not involve a man-month of learning the entire codebase from scratch.

My comment here was suggestions on what you'd need to look at to fix sunrise/sunset based times generally based on what I noticed while fixing the static times case. I never attempted to fix sunrise/sunset based times as it wasn't clear that this was affected by the same issue at the time I wrote the fix.

> > > When I read the test suite, this situation is covered. Yet the tests are not catching it, which points to a problem in the testing implementation not the lack of tests. See my other comment about setting up a working test environment
> 
> > While I haven't used tod with sun events enough to run into this issue, my guess is if it's a similar issue to the PR I fixed it needs some very specific conditions to trigger which may not be the case in the existing unit tests.
> > 
> > * after must be later in the day than "before", ie after: sunset, before: sunrise. There may be even more edge cases where using a mix of sun events and absolute times.
> 
> The current unit tests include one minute after sunset and one minute before sunrise. Both of those should be within the period. Yet in reality the latter time does not work. The test is incorrectly written (using an undocumented mechanism that I gave up trying to figure out)

Yes, but the root cause is that "sunrise" isn't just a time, it's a datetime component. The catch is that the date is today when it should have been yesterday, so "after" is false (since todays sunset hasn't happened yet, but yesterdays sunset has happened which is actually what we were wanting to check against).

> > * home assistant must "start" when you're expecting the state to be on, but in the day after the "after" condition (but still before the "before" time)
> 
> AND after midnight. It works fine if it restarts before midnight.

yes that's what I meant by the "but in the day after" bit. The real edge case here with sunrise/sunset is that it may even be multiple days not just one day.

> That is the main problem with this code -- it's way too "what day am I in?" oriented. Sunset happens at a time. Sunrise happens at a time. Times are integers so it's all simple math. This code needs to stop thinking about days.

Except sunrise and sunset doesn't happen at a time. Sunrise on a particular day may happen at a time, and the next day it'll happen at a different time, and the day after that it'll happen at a different time again. This is the case even when not considering the arctic circle edgecase, which then sunrise happened Monday, then no sun events until the sun finally sets on Thursday.

> > Testing that for example at 1am in a location near the equator I would expect would trigger the issue.
> 
> Testing at 1am ANYWHERE triggers the issue.

The context of this is pointers for writing a unit test to reproduce the issue. When writing a unit test "anywhere" isn't a valid location. Testing too far north or south (ie arctic circle) may be not result in the correct answer being "on" at all times of year, hence suggesting a location near the equator specifically for this case of the unit test.

> > I expect eg test_from_sunset_to_sunrise is not catching this because it is looking up the next sunrise and faking the time as that, so we're effectively in the "one day has passed since restarting home assistant" case there which is when it works.
> 
> The problem is identical to the one you solved. After a restart the sunset event is in the future, not in the past. Forget about days and reset to the sunset prior to the restart.

Simply removing that check causes a whole bunch of unit tests to break. I expect a PR that knowingly causes other issues to fix one issue would not be accepted. Who's to say which issue is more important?

> > I'd write a unit test that specifically hardcodes in times and dates
> 
> Honestly I think that the current code is thinking way too much about specific conditionals around date, location, etc. None of that matters. Reduce Start to an integer, End to an integer, and compare simple math. Anything else is distracting in the code and misleading the reader. With simple integers, it doesn't matter if there are 72 years between sunset and sunrise on a moon of Saturn or whatever, none of the code needs to change 😁

You're free to rewrite the component from scratch if you think you can do it better, just be aware that regressions on existing functionality (ie, as tested by the existing unit tests, unless you can specifically identify a logic flaw in one of the existing unit tests which should be corrected anyway) would be unlikely to be accepted.

jorhett commented:
I'm really confused here by these statements:

@emontnemery said
> please mark as ready for review when you've added a test which fails without your test and passes with it and also tested your patch in a running Home Assistant setup 👍

@nvx said:
> Simply removing that check causes a whole bunch of unit tests to break. I expect a PR that knowingly causes other issues to fix one issue would not be accepted. Who's to say which issue is more important?

Neither of you have taken into the account that the current tests pass and yet the code doesn't work, so the test suite is itself broken. My attempt to read through this code and the test suite shows that both encode the author(s) assumptions rather than test the functionality.  This is why best practice is that implementations and tests are written by different developers -- it flushes out assumptions on both sides.

Like I said, I'm happy to fix code. Happy to write tests. But I won't be able to write a test that supplements currently nonfunctional tests ... there's a bigger problem here.  If you want me to comply with this directive:

> when you've added a test which fails without your test and passes with it 

then I cannot start until you have fixed your tests to correctly report the current (failed) behavior.

jorhett committed:
Fix calculations of ToD when now is between after and before

Remove all sense of "day" from sun events since they are not guaranteed in days
Add get_astral_event_previous() to sun helper
Use get_astral_event_previous() to get previous after event if now < before

jorhett commented:
I don't have tests for you, but I unwinded a troubling number of bugs in the original code and ended up greatly simplifying it.

Bug 1:  incorrect determination of next update
   If now is between After and Before, the current code will set the next update to After when the next event is actually Before https://github.com/home-assistant/core/blob/dev/homeassistant/components/tod/binary_sensor.py#L251-L254
   
Bug 2: this test is backwards.  It says that if Before is prior to After, then revert before to the previous day. In which which both will be prior to the current time 🤦‍♀️  which will mean it will calculate as outside the window and OFF when actually that means it's during a current window and ON https://github.com/home-assistant/core/blob/dev/homeassistant/components/tod/binary_sensor.py#L177-L181

And days have no meaning for sun events which can be multiple days away in certain regions so this is the wrong fix regardless.

Bug 3: offsets were calculated after day-swapping, which can actually create invalid results. Comparison of Before and After must happen with offsets included https://github.com/home-assistant/core/blob/dev/homeassistant/components/tod/binary_sensor.py#L207-L209

The fix added in #65884 was just after-patching this earlier mistake so by fixing the original mistake that scenario should be fixed.

Unless I made a mistake, this should pass all the tests that exist although we know those are broken 🤷‍♀️

emontnemery commented:
> Unless I made a mistake, this should pass all the tests that exist although we know those are broken 🤷‍♀️

Let's improve the tests if they are broken 👍
The code is hairy with a lot of edge cases, and we need good tests covering those edge cases.

KevinCathcart commented:
Looking over this code, the latest committed version does look pretty good.

It is much simpler than before, and far more obvious why the code is (almost) correct. By no longer trying to tie astral events to a date, it avoids the fundamental cause of the bug in question.

I see a new test was added, which looks like it would fail with the previous code. (But I have not verified this). Unfortunately, I have to question if the test code was actually run, and if this code passes all existing tests, given that the test file no longer importing get_astral_event_date means some existing tests would break.

If the new test passes only with the new code, and fails with the old code, and the rest of the tests still pass after fixing the imports, then this should be almost good, although I’d like to see a few more test cases similar to the added one, namely involving starting after midnight inside the time period while using time for after, sun for before, and vice versa.

However, the new code still has bugs. The basic concept is sound. The procedure being determining the next after time and the next before time, and then if the after comes later than the before, change it to the previous after time. I’m pretty sure that I s a correct way of calculating this, and less edge-case full than the previous code. 

However, that is not quite what is currently implemented in the latest commit. There remain two issues. First for time based before and after times, it currently calculates the time for the current day, not the next occurrence of that time (like the sun part does). It needs to add a day to the result of `_naive_time_to_utc_datetime` if that time (adjusted by any offset) is less than nowutc.
Secondly when trying to find the next sun event, you want to subtract the offset from nowutc, otherwise if searching just minutes after sundown, but the target is sundown+1 hour, it will find tomorrow’s sundown, when we really want the one that just happened, since the offset version hasn’t happened yet.

If both of those are addressed, then I’m reasonably confident the algorithm used would be correct, and the important part would be making sure we have the tests that would have caught those two issues in place.

emontnemery commented:
@jorhett If you have not setup a local development environment, you can follow this guide: https://developers.home-assistant.io/docs/development_environment

It will ensure the linters are automatically run when you commit, and it allows you simply run the the `tod` tests from within the Home Assistant venv, for example like this: `pytest -vvx tests/components/tod`

jorhett commented:
> If both of those are addressed, then I’m reasonably confident the algorithm used would be correct, and the important part would be making sure we have the tests that would have caught those two issues in place.

Thank you for the writeup. I'm out of the country on vacation so I doubt I will get back to working on this in the next 5 days. Others are welcome to contribute, or I'll follow through when I return.

jorhett commented:
I finally did get the devcontainer working. PR [#1335](https://github.com/home-assistant/developers.home-assistant/pull/1335) provides a simple startup that works for existing developers.

> If you have not setup a local development environment, you can follow this guide: https://developers.home-assistant.io/docs/development_environment

Running a local development environment on my own (different) hardware, different OS, different Python version is a pointless way to create more work for everyone involved.  It really needs to be documented how to run those tests within the container. Probably exactly the same, but mentioning that would be great.

jorhett commented:
> The instructions need to include `pip3 install -r requirements_test_all.txt`

which doesn't work either. I spent 6 hours hand-cobbling together a working setup that would run pytest. No clue if mine matches anyone else's or not. Seriously, the testing here is worse than a project without tests-- in fact, it defeats the point of having tests at all.  With a good test suite, a new developer is backfilled by tests that confirm their work. In this case I lost my entire evening just trying to get your test suites working and had no more than 20 minutes of working on the code+tests to determine the original testing failures before it was nearly dawn and I needed to get a few hours sleep before work meetings began.

Today, I come back to try and progress further and received an insolent insistence that VScode and Docker are the One True Way and Nothing Else Can Be Supported attitude from the most active HA member.

So your developer setup is incoherent, your test setup instructions don't work, your developer setup doesn't work, a contributor is left to flounder in a non-working environment that only works in some dev setup that they must intuit on their own...   in short, that your test suites pass for code that doesn't work makes perfect sense considering all the other friction points.

Right now, I simply don't see a way to help. I would help fix this code, but I can't be in any way sure that what I hacked together is a realistic or workable test for anything. I would help improve the developer setup instructions, but they are PERFECT AS IS, THANK YOU ... 🤷‍♀️ 

Nabu doesn't want my help, and I'm not going to throw away weeks of effort for something nobody wants.

jorhett commented:
I am finally free of audit season, and my actually have time to work on this soon.

jorhett commented:
I have not forgotten about this PR. But every time I find some hours to work on it, I instead end up fighting setup problems in the repo/tools/docs. So I'm literally burning all my available effort on fixing setup issues, and can't get to a place where I can test my latest changes and be certain they will work in a hass deploy.

jorhett commented:
I have a working but likely inconsistent environment by running the following command by hand:

```shell
$ pip install pytest numpy==1.23.2 sqlalchemy==1.4.44 fnvhash==0.1.0 zeroconf==0.39.4
```

writing this comment here so I can find it next time I have to rebuild the env

jorhett commented:
This is going nowhere. As usual, @frenck has stepped in to be abrasive and insulting. His solution for a dev environment that never works?  Reject any and all steps forward that don't magically intuit the undocumented system.

Fix bugs? Naw.
Improve the developer setup? Naw.
Be as insulting and dismissive as possible, such that possible contributors want nothing to do with the project? Apparently this is his life goal.

frenck commented:
> As usual, @frenck has stepped in to be abrasive and insulting.

I wasn't abrasive, and most certainly, I was never insulting; If I was, please point me out, and I will apologize for it. If not, consider this an official warning from my end as I find these types of acquisitions unacceptable.

You, however, have been quite demanding by expecting me to finish your work, and setting a scene with this response skewing the reality.

You've created a PR without showing any intention of moving it forward. Instead, you said that I should fix it. That is not how PRs work. I'm not here to finish your PRs.

../Frenck

jorhett commented:
> > As usual, @frenck has stepped in to be abrasive and insulting.
> 
> I wasn't abrasive, and most certainly, I was never insulting; If I was, please point me out, and I will apologize for it. If not, consider this an official warning from my end as I find these types of acquisitions unacceptable.

I would be happy to. Is there a better place for this than a PR ?  I'll try to DM you on the forum if that's possible.

frenck reviewed (commented):
Because there hasn't been any activity on this PR for quite some time now, I've decided to close it for being stale.

Feel free to re-open this PR when you are ready to pick up work on it again 👍

../Frenck
